frontend/public/components/routes.tsx

```diff
@@ -200,31 +203,56 @@ const RouteListRow: React.SFC<RoutesRowProps> = ({obj: route}) => <ResourceRow o
 </ResourceRow>;
 
-const TLSSettings = props => <span>
-  {!props.tls && 'TLS is not enabled.'}
-  {props.tls && <dl>
-    <dt>Termination Type</dt>
-    <dd>{props.tls.termination}</dd>
-    <dt>Insecure Traffic</dt>
-    <dd>{props.tls.insecureEdgeTerminationPolicy || '-'}</dd>
-    <dt>Certificate</dt>
-    <dd>{props.tls.certificate || '-'}</dd>
-    <dt>Key</dt>
-    <dd>{props.tls.key || '-'}</dd>
-    <dt>CA Certificate</dt>
-    <dd>{props.tls.caCertificate || '-'}</dd>
-    {_.get(props.tls, 'termination') === 'reencrypt' && <span>
-      <dt>Destination CA Cert</dt>
-      <dd>{props.tls.destinationCACertificate || '-'}</dd>
-    </span>}
-  </dl>
+class TLSSettings extends SafetyFirst<TLSDataProps, TLSDataState> {
+  constructor(props) {
+    super(props);
+    this.state = { showPrivateKey: false };
+    this.toggleKey = this.toggleKey.bind(this);
+  }
+
+  toggleKey() {
+    this.setState({ showPrivateKey: !this.state.showPrivateKey });
+  }
+
+  render() {
+    const { showPrivateKey } = this.state;
+    const { tls } = this.props;
+    const visibleKeyValue = showPrivateKey ? tls.key : <MaskedData /> ;
+
+    return !tls ?
+      'TLS is not enabled.' :
+      <dl>
+        <dt>Termination Type</dt>
+        <dd>{tls.termination}</dd>
+        <dt>Insecure Traffic</dt>
+        <dd>{tls.insecureEdgeTerminationPolicy || '-'}</dd>
+        <dt>Certificate</dt>
+        <dd>{tls.certificate ? <CopyToClipboard value={tls.certificate}/> : '-'}</dd>
+        <dt className="co-m-route-tls-reveal__title">Key {tls.key &&
+          <button className="btn btn-link co-m-route-tls-reveal__btn" type="button" onClick={this.toggleKey}>
+            {
+              showPrivateKey
+                ? <React.Fragment><i className="fa fa-eye-slash" aria-hidden="true"></i> Hide</React.Fragment>
+                : <React.Fragment><i className="fa fa-eye" aria-hidden="true"></i> Reveal</React.Fragment>
+            }
+          </button>}
+        </dt>
+        <dd>{tls.key ? <CopyToClipboard value={tls.key} visibleValue={visibleKeyValue}/> : '-'}</dd>
+        <dt>CA Certificate</dt>
+        <dd>{tls.caCertificate ? <CopyToClipboard value={tls.caCertificate}/> : '-'}</dd>
+        {tls.termination === 'reencrypt' && <React.Fragment>
+          <dt>Destination CA Cert</dt>
+          <dd>{tls.destinationCACertificate ? <CopyToClipboard value={tls.destinationCACertificate}/> : '-'}</dd>
+        </React.Fragment>}
+      </dl>;
   }
-</span>;
+}
 
 const calcTrafficPercentage = (weight: number, route: any) => {
-  let totalWeight = 0;
-  totalWeight += route.spec.to.weight;
-  _.each(route.spec.alternateBackends, alternate => totalWeight += alternate.weight);
+  let totalWeight = _.reduce(route.spec.alternateBackends, (result, alternate) => {
+    return result += alternate.weight;
+  }, 0 + route.spec.to.weight);
+
   return (weight/totalWeight*100).toFixed(1);
```

Comment:
@dtaylor113 Can you confirm it's not possible to set all weights to 0? Otherwise this could be a divide by zero error.

Reply:
Weights can be set to 0, adjusted code to handle